### ISSUE TITLE
Why does {c} return the name of the base class, not the derived class?

### ISSUE THREAD

**Issue by andrewimcclement** [nunit/nunit#4446]
TLDR: `{c}` shows `Base` not `Derived` for the test name.

Using .NET Framework 4.8, NUnit 3.13, NUnit Console Runner 3.16.2.
```csharp
namespace
{
    internal abstract class Base
    {
        [Test]
        public void SomeTest() => Assert.Pass();
    }
    
    internal class Derived1 : Base { }
    internal class Derived2 : Base { }
}
```
I am using `--test-name-format="{C}.{m}{a}"`.

This gives the output of
```
=> Namespace.Derived2.Namespace.Base.SomeTest
=> Namespace.Derived1.Namespace.Base.SomeTest
```
and a TestResults.xml file looking like
```xml
<test-suite type="TestSuite" id="1-23561" name="Namespace" fullname="Namespace" runstate="Runnable" testcasecount="2" result="Passed" start-time="2023-08-01T23:01:05.0870009Z" end-time="2023-08-01T23:01:05.3064107Z" duration="0.214269" total="2" passed="2" failed="0" warnings="0" inconclusive="0" skipped="0" asserts="0">
  <test-suite type="TestFixture" id="1-1000" name="Derived1" fullname="Namespace.Derived1" classname="Namespace.Derived1" runstate="Runnable" testcasecount="1" result="Passed" start-time="2023-08-01T23:01:05.1026235Z" end-time="2023-08-01T23:01:05.3064107Z" duration="0.213245" total="1" passed="1" failed="0" warnings="0" inconclusive="0" skipped="0" asserts="0">
    <properties>
      <property name="Category" value="Special" />
    </properties>
    <test-case id="1-1001" name="Namespace.Base.SomeTest" fullname="Namespace.Derived1.Namespace.Base.SomeTest" methodname="SomeTest" classname="Namespace.Base" runstate="Runnable" seed="302541065" result="Passed" start-time="2023-08-01T23:01:05.1026235Z" end-time="2023-08-01T23:01:05.3064107Z" duration="0.204605" asserts="0">
      <properties />
    </test-case>
  </test-suite>
  <test-suite type="TestFixture" id="1-1002" name="Derived2" fullname="Namespace.Derived2" classname="Namespace.Derived2" runstate="Runnable" testcasecount="1" result="Passed" start-time="2023-08-01T23:01:05.1026235Z" end-time="2023-08-01T23:01:05.3064107Z" duration="0.213246" total="1" passed="1" failed="0" warnings="0" inconclusive="0" skipped="0" asserts="0">
    <properties>
      <property name="Category" value="Special" />
    </properties>
    <test-case id="1-1003" name="Namespace.Base.SomeTest" fullname="Namespace.Derived2.Namespace.Base.SomeTest" methodname="SomeTest" classname="Namespace.Base" runstate="Runnable" seed="95196297" result="Passed" start-time="2023-08-01T23:01:05.1026235Z" end-time="2023-08-01T23:01:05.3064107Z" duration="0.204558" asserts="0">
      <properties />
    </test-case>
  </test-suite>
</test-suite>
```
The name of the test in both cases is `Namespace.Base.SomeTest`, whereas I would like to used the name of the derived class, i.e. to get Namespace.Derived1.SomeTest.  My question is why does this behaviour occur?  If I want the full name of the method (which would be `Namespace.Base.SomeTest` I can use `{M}`.  Currently I can't see a difference between `{M}` and `{C}.{m}`.

For full disclosure I am trying to show the test names in Azure Pipelines, which relies on the test name field rather than the fullname field.  If necessary I can modify the XML directly, but it would be much nicer if NUnit supported the scenario directly.

**Comment by OsirisTerje** [comment]
We need to look into this one.

**Comment by andrewimcclement** [comment]
Started work on a basic fix, though I may need to dump what I have for someone else to pick up.

**Comment by andrewimcclement** [comment]
Opened a PR, no longer actively working on this.  I will do my best to respond to comments when time permits.

**Comment by OsirisTerje** [comment]
@manfred-brands @stevenaw  Comments ?

**Comment by manfred-brands** [comment]
According to the [documentation](https://docs.nunit.org/articles/nunit/running-tests/Template-Based-Test-Naming.html?q=Testname), `{c}` is:
> The class name of the test

To me that means the class the test is defined in, which is here the base class.
Regardless of the derived class, it is the same test. Not two different tests with the same method name.

Having said that, how would one distinguish between the two tests in the derived classes?
Do we need an extra letter?
Otherwise the change would not be backward compatible.

**Comment by andrewimcclement** [comment]
If we insist on backwards compatibility,  then yes, I think we need two new symbols, one to represent the derived class name, one to represent the full name of the derived class.  I guess you could also consider providing the names page of the derived class.

However, I don't like the idea of adding 2-3 new symbols which in the default case are identical to existing symbols.  This is unintuitive and not very helpful for new users.  I would argue our aim should be to provide all the relevant symbols so the user can construct whatever string they want to, hence why in my PR I repurposed the existing class name symbols, as that covers almost all of the possibilities (you can still use `{M}` for the full method name which will include the base class name).

I guess another ugly solution to backwards compatibility is to provide a flag to control which behaviour you get, but that is very ugly.

**Comment by manfred-brands** [comment]
@stevenaw I only learned about these name substitutions a few weeks ago and haven't use them myself. I don't know how much the feature is used.

If the idea is that test names are deterministic, then the derived class name makes more sense. At work we have some base classes that indeed call an abstract method to control behaviour.

**Comment by OsirisTerje** [comment]
I worked with these some years back, but since most of my work was with the adapter in Visual Studio and dotnet test, I found it doesn't make much sense to change these.  The Visual Studio test explorer will figure out the correct naming based on the underlying information, and changing the name itself, which by default is the test itself with its parameters are what the underlying testhost for VS and dotnet expects.

The test explorer will display the code above - using Real Time Test Discovery, with no changes to the name template, as:
![image](https://github.com/nunit/nunit/assets/203432/86f71900-2128-4c92-abb2-7fff6c962e78)
Running the tests give exactly the same structure
![image](https://github.com/nunit/nunit/assets/203432/f04fd22f-c7df-4982-b220-e55efb4dca8c)
Using dotnet test give, with the appropriate options:
![image](https://github.com/nunit/nunit/assets/203432/2105c174-5b89-41e5-ab25-db43e8e13945)

If we change the template though to {C}.{m}{a}:
![image](https://github.com/nunit/nunit/assets/203432/feb1494d-fbe7-4860-98da-b351219fab01)
Those inside the green square is the ones from the RTD, whereas the ones in the yellow square is from the NUnit discovery phase. As can be seen, they are no longer in sync.  The RTD doesn't adhere to the naming templates.  Also, the RTD finds the correct expected naming anyway, whereas the NUnit discovery doesnt. Further, even if it were changed as suggested, it would build the wrong hierarchy, as the full test name would be placed below the class named Derived1/2. 

On dotnet test command line, the naming would be correct without the NUnit option, but you would anyway have to include the runsettings.  It is easier to not do use a changed template but still have correct naming using the DisplayName property instead. 

That said, if you absolutely would like to use the NUnit.Console (not sure why though), a change like this would not do any harm, as long as the defaults are not changed.  It is the {m]{a} setting that is critical to keep.  
I can also be an option to move this functionality over to the NUnit.Console, since the others don't really need this functionality any more.

**Comment by stevenaw** [comment]
Thanks for your insights @OsirisTerje I was going to be asking your thoughts but I see you were too quick for me. I'm generally hesitant to truly suggest any breaking changes to the xml in this area.

@manfred-brands Nor I! Though that we came from similar experience backgrounds yet opposite conclusions to me says a lot about the nature of the problem we're trying to help @andrewimcclement with :)

**Comment by andrewimcclement** [comment]
@OsirisTerje Using the `fullname` property would be ideal for my particular purposes, since all I personally care about is having distinguishable test names.  I think there was some request for Azure Pipelines to provide more flexibility than just using the `name` property - I'll try and dig it up later.  Unfortunately there is no option to use `fullname` when using https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/reference/publish-test-results-v2 for the NUnit format.

As it stands, I would need to add a PowerShell script (or similar) to adjust the XML appropriately, which I'm not keen to maintain if I can avoid it.

The main reason for using NUnit Console Runner over the NUnit test adapter is for speed - our unit tests take ~10 minutes to run using the console runner (around 40k tests), using the test adapter I think it takes ~20 mins, which is rather painful.  (I don't have exact numbers to hand, but the difference was significant enough that we use the console runner everywhere.)

As you say, we definitely don't want to change the default template - test runners inside IDEs already provide significant additional context (i.e. the namespace and class name).  If Azure Pipelines gave the same context, I wouldn't even be bothering to change the default name at all...

@stevenaw I haven't had a chance to try an abstract or virtual test - that definitely seems like a good thing to try, just for comparison purposes.

The point by @manfred-brands that this feature isn't used much is definitely fair.  If the consensus ends up being "Hmm, it could be changed, but it isn't worth the effort" that's definitely a legitimate outcome.  (Though I might open a documentation PR to try and make the current functionality clearer in that case...)

**Comment by andrewimcclement** [comment]
To clarify what happens currently when using abstract tests:

Using
```
namespace Namespace
{
    [Category("hello")]
    internal abstract class Base
    {
        [Test]
        public abstract void SomeTest(); // Please assume this depends on functionality defined in the derived classes.
    }
}

namespace OtherNamespace
{
    using Namespace;

    internal class Derived1 : Base
    {
        [Test]
        public override void SomeTest() => Assert.Pass();
    }

    internal class Derived2 : Base
    {
        [Test]
        public override void SomeTest() => Assert.Pass();
    }
}
```
with command ` ~/.nuget/packages/nunit.consolerunner/3.16.3/tools/nunit3-console.exe Random.UnitTests.dll --where="cat == hello" --test-name-format="{C}.{m}{a}"` I get:

```
<test-suite type="TestSuite" id="1-23582" name="OtherNamespace" fullname="OtherNamespace" runstate="Runnable" testcasecount="2" result="Passed" start-time="2023-10-06T14:08:22.6764572Z" end-time="2023-10-06T14:08:22.8244590Z" duration="0.147668" total="2" passed="2" failed="0" warnings="0" inconclusive="0" skipped="0" asserts="0">
  <test-suite type="TestFixture" id="1-1002" name="Derived2" fullname="OtherNamespace.Derived2" classname="OtherNamespace.Derived2" runstate="Runnable" testcasecount="1" result="Passed" start-time="2023-10-06T14:08:22.6774554Z" end-time="2023-10-06T14:08:22.8234588Z" duration="0.145950" total="1" passed="1" failed="0" warnings="0" inconclusive="0" skipped="0" asserts="0">
    <properties>
      <property name="Category" value="hello" />
    </properties>
    <test-case id="1-1003" name="OtherNamespace.Derived2.SomeTest" fullname="OtherNamespace.Derived2.OtherNamespace.Derived2.SomeTest" methodname="SomeTest" classname="OtherNamespace.Derived2" runstate="Runnable" seed="776068731" result="Passed" start-time="2023-10-06T14:08:22.6824509Z" end-time="2023-10-06T14:08:22.8184567Z" duration="0.136289" asserts="0">
      <properties />
    </test-case>
  </test-suite>
  <test-suite type="TestFixture" id="1-1000" name="Derived1" fullname="OtherNamespace.Derived1" classname="OtherNamespace.Derived1" runstate="Runnable" testcasecount="1" result="Passed" start-time="2023-10-06T14:08:22.6774554Z" end-time="2023-10-06T14:08:22.8234588Z" duration="0.146026" total="1" passed="1" failed="0" warnings="0" inconclusive="0" skipped="0" asserts="0">
    <properties>
      <property name="Category" value="hello" />
    </properties>
    <test-case id="1-1001" name="OtherNamespace.Derived1.SomeTest" fullname="OtherNamespace.Derived1.OtherNamespace.Derived1.SomeTest" methodname="SomeTest" classname="OtherNamespace.Derived1" runstate="Runnable" seed="225199244" result="Passed" start-time="2023-10-06T14:08:22.6824509Z" end-time="2023-10-06T14:08:22.8184567Z" duration="0.136385" asserts="0">
      <properties />
    </test-case>
  </test-suite>
</test-suite>
```

**Comment by stevenaw** [comment]
Thanks for testing that out @andrewimcclement . I recognize that it may mean some refactoring but would making the method abstract be another option for you to get the output you need?

**Comment by andrewimcclement** [comment]
Making methods abstract (but with exactly the same implementation) doesn't aid code reuse, which I care about more than this reporting issue.

Wrt the type name, certainly, from the perspective of the method, it seems perfectly logical.  I'm not arguing for a change to `{m}` or `{M}` which should exactly match `methodInfo.Name` and `$"{methodInfo.DeclaringType.FullName}.{methodInfo.Name}"`.  My argument is around whether we should have `{c}` as `methodInfo.DeclaringType` or `typeof(MyFixture)`, of which I find the latter far more useful.  (The symbols we make available for use and their meaning is definitely an nunit issue.)

**Comment by OsirisTerje** [comment]
To avoid making any backward issues coming up, wouldn't it be easiest to just add another letter to indicate we want the name of the most derived class, the class that is actually being instantiated?

**Comment by stevenaw** [comment]
Could you provide an example a bit more representative to how your tests are structured? It would help us to understand any polymorphic relationships, particularly in helping to think through options which may help you get the output you want while maximizing reuse without having to write a custom script or wait for any desired change in a future release. 

FWIW, I'm just one team member but I am leaning towards any change here being in the form of a new symbol. NUnit doesn't have any internal knowledge of test structure and so we must rely on what the C# compiler makes available to us via reflection based on how the test methods are defined, and to support a new behaviour it may be safest to add a new symbol

**Comment by stevenaw** [comment]
I see @OsirisTerje replied while I had been typing, thanks!

**Comment by OsirisTerje** [comment]
And, I see @stevenaw  replied while I had been typing, thanks!

But it seems we do agree, so can we then ask @andrewimcclement  if he wants to contribute a PR?

**Comment by stevenaw** [comment]
I think @manfred-brands can also be counted in the agreement too based on an earlier comment.

@OsirisTerje @manfred-brands did you have any thoughts on what letter/symbol may make sense to add?

@andrewimcclement would you be interested in updating your PR based on the discussion here?

**Comment by OsirisTerje** [comment]
It could be {d} for derived class name, or {e} or {x] for "exact class",  but really any unused character ref, https://docs.nunit.org/articles/nunit/running-tests/Template-Based-Test-Naming.html 
Any one is good with me.

**Comment by stevenaw** [comment]
No strong preferences from me either.
I'm partial to `{d}` out of those options myself as it feels more natural to me, but I'm also fine with anything.

**Comment by andrewimcclement** [comment]
I agree with {d} (and {D}).  Got a branch but having trouble actually running the tests in TestNameGeneratorTests.cs - some guidance would be welcome.

**Comment by andrewimcclement** [comment]
Apologies for the delay, life happened.
